lib/slack-hook-handler.js

```diff
@@ -31,22 +33,120 @@ SlackHookHandler.prototype.getIntent = function(slackUser) {
  * @param {Object} params HTTP body of the webhook request, as a JSON-parsed dictionary.
  */
 SlackHookHandler.prototype.handle = function(params) {
-    if (params.user_id !== "USLACKBOT" && params.text) {
-        var intent = this.getIntent(params.user_name);
-        if (this.rooms.knowsSlackChannel(params.channel_id)) {
-            var roomID = this.rooms.matrixRoomID(params.channel_id);
-            if (roomID) {
-                intent.sendText(roomID, substitutions.slackToMatrix(params.text));
+    if (params.user_id === "USLACKBOT") {
+        return;
+    }
+    if (!this.rooms.knowsSlackChannel(params.channel_id)) {
+        console.log(
+            "Ignoring message for slack channel " +
+            "with unknown matrix ID: " + params.channel_id +
+            " (" + params.channel_name + ")"
+        );
+        return;
+    }
+    var intent = this.getIntent(params.user_name);
+    var roomID = this.rooms.matrixRoomID(params.channel_id);
+    if (params.text) {
+        intent.sendText(roomID, substitutions.slackToMatrix(params.text));
+        return;
+    }
+    this.lookupAndSendNonTextMessage(params.channel_id, params.timestamp, intent, roomID);
+    return;
+};
+
+/**
+ * Attempts to handle a non-text message received from a slack webhook request.
+ *
+ * @param {string} channelID Slack channel ID.
+ * @param {string} timestamp Timestamp when message was received, in seconds
+ *     formatted as a float.
+ * @param {Intent} intent Intent for sending messages as the relevant user.
+ * @param {string} roomID Matrix room ID associated with channelID.
+ */
+SlackHookHandler.prototype.lookupAndSendNonTextMessage =
+        function(channelID, timestamp, intent, roomID) {
+    if (!this.config.slack_master_token) {
+        return;
+    }
+    var params = {
+        channel: channelID,
+        latest: timestamp,
+        oldest: timestamp,
+        inclusive: "1",
+        token: this.config.slack_master_token,
+    };
+    var self = this;
+    this.requestLib.post("https://slack.com/api/channels.history", {form: params},
+            function(err, _, body) {
+        if (err) {
+            console.log("Error looking up history: %s", err);
+        }
+        else {
+            var resp = JSON.parse(body);
+            if (!resp.ok || !resp.messages) {
+                console.log("Could not find history: " + body);
+                return;
+            }
+            if (resp.messages.length != 1) {
+                console.log("Really unlucky, got multiple messages at same" +
+                    " microsecond, dropping:" + body);
```

Comment:
A bit of explanation here would be good. I understand we're dropping something because `messages.length > 1` (which is only because of the null check earlier, it may be clearer to say `if (resp.messages.length > 1)` instead) but it's not clear that this is due to a race with them having the same `timestamp` value. It would probably also be polite to log the things being dropped, rather than just the raw response and let the poor person looking at the logs to assume/have to look at the code to find out that only the first element is being sent.